### PR TITLE
Handle non blocking private chains

### PR DESCRIPTION
**Pull request by djlmills**
These changes make the middleware more robust. It will now start the chain sponsorship process to create a new private chain if the local chain does not block within a short timeout. This means it can now recover when the back-end node is restarted and the browser chain in local storage is now invalid.

### CODE REVIEW

File: packages/interbit-ui-tools/src/middleware/middleware.js [hunk 1]
```diff
@@ -105,15 +121,9 @@ const subscribeToChain = (store, chainAlias, chain) => {
   chain.blockSubscribe(() => {
     const newBlock = chain.getCurrentBlock()
     store.dispatch(actionCreators.chainBlockAdded(chainAlias, newBlock))
-
-    if (firstBlock) {
-      store.dispatch(actionCreators.chainBlocking(chainAlias))
-      firstBlock = false
-    }
```
Comment (djlmills) [inline packages/interbit-ui-tools/src/middleware/middleware.js]: Now handled by block detection in chains.js

File: packages/interbit-ui-tools/src/middleware/sagas/chains.js [hunk 2]
```diff
@@ -208,25 +217,21 @@ function* loadChain({ cli, chainAlias, chainId }) {
 
   yield put.resolve(actionCreators.chainLoaded({ chainAlias, chainId }))
 
-  return chain
+  yield call(detectBlocking, chain)
+
+  yield put.resolve(actionCreators.chainBlocking({ chainAlias }))
```
Comment (djlmills) [inline packages/interbit-ui-tools/src/middleware/sagas/chains.js]: Wait for chain to start blocking, then dispatch the chainBlocking action. detectBlocking will throw if the chain does not start blocking.

File: packages/interbit-ui-tools/src/middleware/sagas/chains.js [hunk 3]
```diff
@@ -247,7 +252,22 @@ function* tryDeleteChain({ cli, chainAlias, chainId }) {
   }
 }
 
+const detectBlocking = (chain, maxTime = 5000) =>
```
Comment (djlmills) [inline packages/interbit-ui-tools/src/middleware/sagas/chains.js]: This is the new block detection logic.

File: packages/interbit-ui-tools/src/middleware/sagas/index.js [hunk 4]
```diff
@@ -2,24 +2,25 @@ const { put, takeEvery, call } = require('redux-saga/effects')
...
-  )
+  console.log(`${LOG_PREFIX}: *rootSaga(): listening for: `, [
+    actionTypes.STATIC_CHAINS_SAGA,
+    actionTypes.PRIVATE_CHAIN_SAGA,
+    actionTypes.SPONSOR_CHAIN_SAGA
+  ])
+  yield takeEvery(actionTypes.STATIC_CHAINS_SAGA, staticChainsSaga)
   yield takeEvery(actionTypes.PRIVATE_CHAIN_SAGA, privateChainSaga)
+  yield takeEvery(actionTypes.SPONSOR_CHAIN_SAGA, sponsorChainSaga)
+
```
Comment (djlmills) [inline packages/interbit-ui-tools/src/middleware/sagas/index.js]: The saga will wait for any of these saga actions.

File: packages/interbit-ui-tools/src/middleware/sagas/index.js [hunk 5]
```diff
@@ -2,24 +2,25 @@ const { put, takeEvery, call } = require('redux-saga/effects')
...
+  yield put(actionCreators.staticChainsSaga())
 }
```
Comment (djlmills) [inline packages/interbit-ui-tools/src/middleware/sagas/index.js]: This starts the saga to load the static chains. We know that the preconditions are ready.

File: packages/interbit-ui-tools/src/middleware/sagas/index.js [hunk 6]
```diff
@@ -29,17 +30,9 @@ function* loadInterbitSaga() {
     const { cli } = yield call(interbitContext)
 
     const config = yield call(getHtmlConfig, document)
-    yield put(actionCreators.initialConfig(config))
+    yield put.resolve(actionCreators.initialConfig(config))
 
-    const peers = getConfiguredPeers(config, { subtree: entireTree })
-    yield call(connectToPeers, { cli, peers })
-
-    const chainsToLoad = getConfiguredChains(config, {
```
Comment (djlmills) [inline packages/interbit-ui-tools/src/middleware/sagas/index.js]: getConfiguredPeers / getConfiguredChains now read from state where needed, simplifying this

File: packages/interbit-ui-tools/src/middleware/sagas/index.js [hunk 7]
```diff
@@ -29,17 +30,9 @@ function* loadInterbitSaga() {
...
-      subtree: entireTree
-    })
-    for (const [chainAlias, { chainId }] of Object.entries(chainsToLoad)) {
-      yield call(loadChain, { cli, chainAlias, chainId })
-    }
```
Comment (djlmills) [inline packages/interbit-ui-tools/src/middleware/sagas/index.js]: Loading static chains now handled by staticChainsSaga

File: packages/interbit-ui-tools/src/middleware/sagas/chains.js [hunk 8]
```diff
@@ -1,10 +1,23 @@
...
+
+  console.log(`${LOG_PREFIX}: *loadStaticChains()`, chainsToLoad)
+
+  const loadChainCalls = Object.entries(chainsToLoad).reduce(
+    (acc, [chainAlias, { chainId }]) =>
+      acc.concat(call(tryLoadChain, { cli, chainAlias, chainId })),
+    []
+  )
+  yield all(loadChainCalls)
+}
```
Comment (djlmills) [inline packages/interbit-ui-tools/src/middleware/sagas/chains.js]: yield all means the chains are loaded in parallel.

File: packages/interbit-ui-tools/src/middleware/selectors.js [hunk 9]
```diff
@@ -40,7 +37,7 @@ const isPublicChainLoaded = (
 ) => isChainLoaded(state, { subtree, chainAlias: publicChainAlias })
 
 const getChainId = (state, { subtree = interbitSubtree, chainAlias } = {}) =>
-  subtree(state).getIn([CHAINS, chainAlias, INTERBIT, CHAIN_ID])
+  subtree(state).getIn([CHAINS, chainAlias, ...INTERBIT_PATHS.CHAIN_ID])
```
Comment (nicholeous) [inline packages/interbit-ui-tools/src/middleware/selectors.js]: :+1:

**Pull request merged**